### ISSUE TITLE
default filter is throwing TypeError: Cannot read property 'toString' of undefined

### ISSUE THREAD

**Issue by dyoung522** [gregnb/mui-datatables#299]
The following code snippet displays the DataTable properly, but then throws `TypeError: Cannot read property 'toString' of undefined` when the 'filter' icon is clicked.

Not sure if this is a bug, or I'm missing something.

```
    const tableOptions = {
      filter: true,
      pagination: false,
      download: false,
      selectableRows: false,
      print: false,
      viewColumns: false
    }

    const charColumns = [
      { name: 'ID', options: { display: false, filter: false, sort: false, download: true } },
      { name: 'Name', options: { display: true, filter: true, sort: true, download: true } },
      { name: 'Level', options: { display: true, filter: true, sort: true, download: true } },
      { name: 'Race', options: { display: true, filter: true, sort: true, download: true } },
      { name: 'Class', options: { display: true, filter: true, sort: true, download: true } },
      { name: 'Sex', options: { display: true, filter: true, sort: true, download: true } },
      { name: 'Owner', options: { display: true, filter: true, sort: true, download: true } }
    ]

    // Creates an array of arrays from Firestore data
    const charData =
      isLoaded(characters) && isLoaded(users)
        ? Object.keys(characters).map(id => {
          const char = characters[id]

          return isEmpty(char)
            ? []
            : [id, char.name, char.level, char.race, char.class, char.sex, users[char.owner].displayName]
        }).sort((a, b) => a[1] - b[1])
        : [['loading', 'Loading... please wait']]

    return (
        <Paper className={classes.root}>
          <MUIDataTable title={'Guild Roster'} data={charData} columns={charColumns} options={tableOptions} />
        </Paper>
    )
```
Here's the raw `charData`:

```
[Array(7)]
0: (7) ["5GM0mnQs2xZP1zCms9xK", "Squall", "5", "Tabaxi", "Monk", "Male", "Donovan C. Young"]
```

## Your Environment

| Tech         | Version |
|--------------|---------|
| Material-UI  |  3.6.0  |
| MUI-datatables  |   2.0.0-beta-41 |
| React        |   16.6.3  |
| browser      |  Chrome  |
| etc          | react-redux-firebase v.2.2.4 |

**Comment by dyoung522** [comment]
Interestingly, if I use the raw `charData` array (hardcoded) it works. So there must be something about the way I'm creating the `charData` above, but I haven't see any differences in the console. Both seem to produce the identical array or arrays.

**Comment by hwatersiv** [comment]
This error also occurs with 'null' values

**Comment by zwk5004** [comment]
Hitting the same issue as well with undefined

**Comment by ideoclickVanessa** [comment]
I keep getting distracted from making a PR; however, for at least the non-remote version of `Uncaught TypeError: Cannot convert undefined or null to object` I believe the following lines in TableFilter.js are the offenders:
[Line 176](https://github.com/gregnb/mui-datatables/blob/master/src/components/TableFilter.js#L176)
[Line 210](https://github.com/gregnb/mui-datatables/blob/master/src/components/TableFilter.js#L210)
[Line 265](https://github.com/gregnb/mui-datatables/blob/master/src/components/TableFilter.js#L265)

**Comment by aprakash-sovrn** [comment]
I believe I'm running into this as well, is anyone aware of a resolution or workarounds?

**Comment by chennanoka** [comment]
finally solved the problem by filling empty strings into the data array when the actual data is not retrieved from the server yet. for us, the bug happens when the "data" is [[],[]] but the 'columns' has value e.g ['column1','column2']. In this case, it tries to get row data from []. instead you need [['',''],['','']].

**Comment by gabrielliwerant** [comment]
@dyoung522 I attempted to recreate the issue here https://3x8wnn3nxp.codesandbox.io/, using `setTimeout` to mimic a response from a server. I do not experience the issue with this example, so there may be something more nuanced with what you're doing, or the issue might have been fixed in the mean time.

Can you take a look and see if you can break what I did or see if you can integrate my approach into your code and let me know how that goes?

**Comment by gabrielliwerant** [comment]
Closing this issue as it seems to be fixed and there hasn't been any new activity here in a while. If the issue crops up again, please open a new issue with a reproducible example.

**Comment by ericpyle** [comment]
@gabrielliwerant I ran into this again with 2.3.0. The workaround is to pad the data empty strings rather than undefined values. You can reproduce the error by clicking on the filter button (configured for multiselect). https://codesandbox.io/embed/muidatatables-crash-multiselect-filters-undefined-data-fnfbe

**Comment by gabrielliwerant** [comment]
@ericpyle Thank you, you're right, `undefined` was an edge case that was not being covered. New fix available here https://github.com/gregnb/mui-datatables/pull/669.

**Comment by gabrielliwerant** [comment]
Should be fixed again in `2.4.0`.
